src/core/api.rs

```diff
@@ -174,99 +178,105 @@ impl Covercrypt {
+    ) -> Result<Vec<u8>, Error>;
 
-    /// Encrypts the given plaintext using the given symmetric key.
-    ///
-    /// The encryption scheme used is AES-256 GCM.
-    ///
-    /// - `symmetric_key`   : AES key
-    /// - `plaintext`       : data to be encrypted
-    /// - `ad`              : optional associated data
-    pub fn dem_encrypt(
-        &self,
-        symmetric_key: &SymmetricKey<SEED_LENGTH>,
-        plaintext: &[u8],
+    fn decrypt(key: &Secret<KEY_LENGTH>, ctx: &[u8], ad: Option<&[u8]>) -> Result<Vec<u8>, Error>;
+}
+
+impl AE<{ Self::KEY_LENGTH }, { Self::NONCE_LENGTH }, { Self::MAC_LENGTH }> for Aes256Gcm {
+    const KEY_LENGTH: usize = Aes256Gcm::KEY_LENGTH;
+    const MAC_LENGTH: usize = Aes256Gcm::MAC_LENGTH;
+    const NONCE_LENGTH: usize = Aes256Gcm::NONCE_LENGTH;
+    fn encrypt(
+        key: &SymmetricKey<{ Self::KEY_LENGTH }>,
+        ptx: &[u8],
         ad: Option<&[u8]>,
+        rng: &mut impl CryptoRngCore,
     ) -> Result<Vec<u8>, Error> {
-        let aes256gcm = Aes256Gcm::new(symmetric_key);
-        let nonce = Nonce::new(&mut *self.rng.lock().expect("could not lock mutex"));
-        let mut ciphertext = aes256gcm.encrypt(&nonce, plaintext, ad)?;
-        let mut res =
-            Vec::with_capacity(plaintext.len() + Aes256Gcm::MAC_LENGTH + Aes256Gcm::NONCE_LENGTH);
-        res.extend(nonce.0);
+        let nonce = &self::Nonce::<{ Self::NONCE_LENGTH }>::new(&mut *rng);
+        let aes = Self::new(key);
+        let mut ciphertext = aes
+            .encrypt(nonce, ptx, ad)
+            .map_err(Error::CryptoCoreError)?;
+        let mut res = Vec::with_capacity(ptx.len() + Self::MAC_LENGTH + Self::NONCE_LENGTH);
+        res.extend(&Nonce::to_bytes(nonce));
         res.append(&mut ciphertext);
         Ok(res)
     }
 
-    /// Decrypts the given ciphertext using the given symmetric key.
-    ///
-    /// The encryption scheme used is AES-256 GCM.
-    ///
-    /// - `symmetric_key`   : AES key
-    /// - `ciphertext`      : encrypted data
-    /// - `ad`              : associated data
-    pub fn dem_decrypt(
-        &self,
-        symmetric_key: &SymmetricKey<SEED_LENGTH>,
-        ciphertext: &[u8],
+    fn decrypt(
+        key: &Secret<{ Self::KEY_LENGTH }>,
+        ctx: &[u8],
         ad: Option<&[u8]>,
     ) -> Result<Vec<u8>, Error> {
-        let aes256gcm = Aes256Gcm::new(symmetric_key);
-        aes256gcm
-            .decrypt(
-                &Nonce::try_from_slice(&ciphertext[..Aes256Gcm::NONCE_LENGTH])?,
-                &ciphertext[Aes256Gcm::NONCE_LENGTH..],
-                ad,
-            )
+        if ctx.len() < Self::NONCE_LENGTH {
+            return Err(Error::CryptoCoreError(
+                cosmian_crypto_core::CryptoCoreError::DecryptionError,
+            ));
+        }
+        let sym_key = SymmetricKey::try_from_slice(key)?;
+        let nonce =
+            &Nonce::<{ Self::NONCE_LENGTH }>::try_from_slice(&ctx[..{ Self::NONCE_LENGTH }])?;
```

Comment:
These curly braces should not be needed, the nonce length can be inferred and avoid taking only refs on your variables (it already tricked you into unnecessarily cloning).
```suggestion
            Nonce::try_from_slice(&ctx[..Self::NONCE_LENGTH])?;
```